### PR TITLE
separate sigcontext type from cpu state type

### PR DESCRIPTION
**Pull request by stsp**
First pass.

With this, the dosemu2 core should no longer depend on a sigcontext type, which should make it finally portable.

### CODE REVIEW

File: src/dosext/dpmi/dnative.c [hunk 1]
```diff
@@ -59,6 +59,10 @@ static void copy_context(sigcontext_t *d, sigcontext_t *s)
 
 static void copy_to_dpmi(sigcontext_t *scp, cpuctx_t *s)
 {
+#ifdef __x86_64__
+  /* needs to clear high part of rip or sighandler stops giving err codes :( */
+  scp->rip = 0;
```
Comment (stsp) [inline src/dosext/dpmi/dnative.c:65]: Hi @amluto I am hitting something very
strange here. I've found that, when returning
to compatibility mode of x86_64 via sigreturn(),
you need to clear the high dword of a RIP
register. If you don't do that, the sighandler
no longer sets the valid error codes for GPF,
even though the trapno is set properly.
Do you have any idea what's going on here?
I thought high parts of 64bit regs are completely
irrelevant in compatibility mode, so why should
I clear RIP?

**Comment by stsp** [comment]
Looks like kernel have seriously
regressed since the last time I played
with native sighandling.
Lets see if lkml can offer some help:
https://marc.info/?l=linux-kernel&m=167066330618354&w=2

**Comment by stsp** [comment]
So according to github stats and
linux kernel log, @amluto is completely
inactive. What happened? :(
And who's going to maintain the sigreturn
code he added to linux kernel for us?
Well, H Peter Anvin and Eric Biederman
(@hpax @ebiederm @ebiederman @ebiederman8)
look like the active contributors to kernel's
signal handling subsystem, so maybe they
may be interested in looking into the bugs
I described here:
https://marc.info/?l=linux-kernel&m=167066330618354&w=2

**Comment by stsp** [comment]
So with some help from LKML
(thanks Eric, @amluto and Thomas
Gleixner), I seem to be coming to the
conclusion that AMD CPU has a bug
that results in checking RIP against CS
limit when returning to compatibility mode.

Their pseudo-code seem to support that:
```
IF ((64BIT_MODE) || (changing_CPL))
    POP.v temp_RSP // in 64-bit mode or changing CPL, IRET always pops SS:RSP
    POP.v temp_SS
CS = READ_DESCRIPTOR (temp_CS, iret_chk)
IF ((64BIT_MODE) && (temp_RIP is non-canonical) ||
    (!64BIT_MODE) && (temp_RIP > CS.limit))
        EXCEPTION [#GP(0)
```
Whereas Intel pseudo-code looks like this:
```
RETURN-TO-OUTER-PRIVILEGE-LEVEL:
    IF OperandSize = 32
            THEN
                        tempESP := Pop();
                        tempSS := Pop(); (* 32-bit pop, high-order 16 bits discarded *)
    ELSE IF OperandSize = 16
            THEN
                        tempESP := Pop(); (* 16-bit pop; clear upper bits *)
                        tempSS := Pop(); (* 16-bit pop *)
            ELSE (* OperandSize = 64 *)
                        tempRSP := Pop();
                        tempSS := Pop(); (* 64-bit pop, high-order 48 bits discarded *)
    FI;
    IF new mode ≠ 64-Bit Mode
            THEN
                        IF EIP is not within CS limit
                                THEN #GP(0); FI;
            ELSE (* new mode = 64-bit mode *)
                        IF RIP is non-canonical
                                    THEN #GP(0); FI;
```
... which seems perfectly correct.

**Comment by stsp** [comment]
[sigreturn.c.txt](https://github.com/dosemu2/dosemu2/files/10217371/sigreturn.c.txt)
Can someone with Intel CPU to
please compile this test-case:
```
$ cc -o sigreturn_64 -no-pie sigreturn.c
$ ./sigreturn_64
```
and copy-paste the last line of the output?

**Comment by stsp** [comment]
@ebiederm Eric, btw all e-mails CCed
to you, got a delivery failure.

**Comment by andrewbird** [comment]
I needed to change the include to the one in my kernel srcs.
~~~
$ lscpu
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   39 bits physical, 48 bits virtual
CPU(s):                          8
On-line CPU(s) list:             0-7
Thread(s) per core:              2
Core(s) per socket:              4
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           140
Model name:                      11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
Stepping:                        1
CPU MHz:                         2400.000
CPU max MHz:                     4200.0000
CPU min MHz:                     400.0000
BogoMIPS:                        4838.40
Virtualisation:                  VT-x
L1d cache:                       192 KiB
L1i cache:                       128 KiB
L2 cache:                        5 MiB
L3 cache:                        8 MiB
NUMA node0 CPU(s):               0-7
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled v
                                 ia prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user
                                  pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RS
                                 B filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtr
                                 r pge mca cmov pat pse36 clflush dts acpi mmx f
                                 xsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rd
                                 tscp lm constant_tsc art arch_perfmon pebs bts 
                                 rep_good nopl xtopology nonstop_tsc cpuid aperf
                                 mperf tsc_known_freq pni pclmulqdq dtes64 monit
                                 or ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr 
                                 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc
                                 _deadline_timer aes xsave avx f16c rdrand lahf_
                                 lm abm 3dnowprefetch cpuid_fault epb cat_l2 inv
                                 pcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_en
                                 hanced tpr_shadow vnmi flexpriority ept vpid ep
                                 t_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 er
                                 ms invpcid rdt_a avx512f avx512dq rdseed adx sm
                                 ap avx512ifma clflushopt clwb intel_pt avx512cd
                                  sha_ni avx512bw avx512vl xsaveopt xsavec xgetb
                                 v1 xsaves split_lock_detect dtherm ida arat pln
                                  pts hwp hwp_notify hwp_act_window hwp_epp hwp_
                                 pkg_req avx512vbmi umip pku ospke avx512_vbmi2 
                                 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg 
                                 avx512_vpopcntdq rdpid movdiri movdir64b fsrm a
                                 vx512_vp2intersect md_clear flush_l1d arch_capa
                                 bilities
~~~

~~~
ajb@sophie:~$ locate desc_defs.h
/usr/src/linux-hwe-5.15-headers-5.15.0-53/arch/x86/include/asm/desc_defs.h
/usr/src/linux-hwe-5.15-headers-5.15.0-56/arch/x86/include/asm/desc_defs.h
ajb@sophie:~$ cc -o sigreturn_64 -no-pie sigreturn.c
ajb@sophie:~$ ./sigreturn_64
[OK]	set_thread_area refused 16-bit data
[OK]	set_thread_area refused 16-bit data
[RUN]	Valid sigreturn: 64-bit CS (33), 32-bit SS (2b, GDT)
[OK]	all registers okay
[RUN]	Valid sigreturn: 32-bit CS (23), 32-bit SS (2b, GDT)
[NOTE]	SP: 8badf00d5aadc0de -> 5aadc0de
[OK]	all registers okay
[RUN]	Valid sigreturn: 16-bit CS (37), 32-bit SS (2b, GDT)
err=18a trapno=d si_code=80 ip=3
~~~

**Comment by stsp** [comment]
> err=18a trapno=d si_code=80 ip=3

Thanks!
The AMD output is:
`err=0 trapno=d si_code=80 ip=100000003`
So indeed its an AMD bug.
But since it conforms to their manual,
all I need is to clear the high parts of
RIP and RSP, and things will work.

**Comment by hpax** [comment]
On December 10, 2022 10:21:15 PM PST, Stas Sergeev ***@***.***> wrote:
>So according to github stats and
>linux kernel log, @amluto is completely
>inactive. What happened? :(
>And who's going to maintain the sigreturn
>code he added to linux kernel for us?
>Well, H Peter Anvin and Eric Biederman
>***@***.*** @ebiedem @ebiederman @ebiederman8)
>look like the active contributors to kernel's
>signal handling subsystem, so maybe they
>may be interested in looking into the bugs
>I described here:
>https://marc.info/?l=linux-kernel&m=167066330618354&w=2
>

@amluto I believe is on paternal leave atm.

If you poke me specifically I can certainly help.

**Comment by stsp** [comment]
Thanks @hpax !
I'll note that for the future.
As for this particular problem -
there appears to be no regression
here. It just appeared that AMD
CPU upon iret to 32bit, checks
full RIP and RSP values against
CS and SS 32bit limits, even if
that sounds silly. And as the result,
I was getting GPF when high part
of RIP was not cleared. Intel CPU
do not do such a thing, so I suspected
a regression, but it was not the case.

Maybe you want to clear high parts
of such regs in sigreturn() to mitigate
such a difference between Intel and AMD
CPUs. Or maybe you want to take a
look at why under GDB the sighandler
prints wrong IP value when coming
from 32 to 64bits. But these tasks are
not important to dosemu2 and I don't
know if you want to experiment with
that just for overall completeness.
If you do, here is the test-case:
https://github.com/dosemu2/dosemu2/files/10217371/sigreturn.c.txt

**Comment by hpax** [comment]
On December 13, 2022 8:36:39 AM PST, Stas Sergeev ***@***.***> wrote:
>Thanks @hpax !
>I'll note that for the future.
>As for this particular problem -
>there appears to be no regression
>here. It just appeared that AMD
>CPU upon iret to 32bit, checks
>full RIP and RSP values against
>CS and SS 32bit limits, even if
>that sounds silly. And as the result,
>I was getting GPF when high part
>of RIP was not cleared. Intel CPU
>do not do such a thing, so I suspected
>a regression, but it was not the case.
>
>Maybe you want to clear high parts
>of such regs in sigreturn() to mitigate
>such a difference between Intel and AMD
>CPUs. Or maybe you want to take a
>look at why under GDB the sighandler
>prints wrong IP value when coming
>from 32 to 64bits. But these tasks are
>not important to dosemu2 and I don't
>know if you want to experiment with
>that just for overall completeness.
>If you do, here is the test-case:
>https://github.com/dosemu2/dosemu2/files/10217371/sigreturn.c.txt
>

For AMD issues, especially non-urgent ones, Borislav Petkov would probably be your preferred contact.

**Pull request merged**